iroh/src/commands/repl.rs

```diff
@@ -0,0 +1,168 @@
+    Exit,
+}
+
+pub struct Repl {
+    env: ConsoleEnv,
+    cmd_tx: mpsc::Sender<(ReplCmd, oneshot::Sender<ToRepl>)>,
+}
+impl Repl {
+    pub fn spawn(env: ConsoleEnv) -> mpsc::Receiver<(ReplCmd, oneshot::Sender<ToRepl>)> {
+        let (cmd_tx, cmd_rx) = mpsc::channel(1);
+        let repl = Repl { env, cmd_tx };
+        std::thread::spawn(move || {
+            if let Err(err) = repl.run() {
+                println!("> repl crashed: {err}");
+            }
+        });
+        cmd_rx
+    }
+    pub fn run(self) -> anyhow::Result<()> {
+        let mut rl =
+            DefaultEditor::with_config(Config::builder().check_cursor_position(true).build())?;
+        let history_path = ConsolePaths::History.with_env()?;
+        rl.load_history(&history_path).ok();
+        loop {
+            // prepare a channel to receive a signal from the main thread when a command completed
+            let (reply_tx, reply_rx) = oneshot::channel();
+            let readline = rl.readline(&self.prompt());
+            match readline {
+                Ok(line) if line.is_empty() => continue,
+                Ok(line) => {
+                    rl.add_history_entry(line.as_str())?;
+                    let cmd = parse_cmd::<ReplCmd>(&line);
+                    if let Some(cmd) = cmd {
+                        self.cmd_tx.blocking_send((cmd, reply_tx))?;
+                    } else {
+                        continue;
+                    }
+                }
+                Err(ReadlineError::Interrupted | ReadlineError::Eof) => {
+                    self.cmd_tx.blocking_send((ReplCmd::Exit, reply_tx))?;
+                }
+                Err(ReadlineError::WindowResized) => continue,
+                Err(err) => return Err(err.into()),
+            }
+            // wait for reply from main thread
+            match reply_rx.blocking_recv()? {
+                ToRepl::Continue => continue,
+                ToRepl::Exit => break,
+            }
+        }
+        rl.save_history(&history_path).ok();
+        Ok(())
+    }
+
+    pub fn prompt(&self) -> String {
+        let mut pwd = String::new();
+        if let Some(author) = &self.env.author(None).ok() {
+            pwd.push_str(&format!(
+                "{}{} ",
+                "author:".blue(),
+                fmt_short(author.as_bytes()).blue().bold(),
+            ));
+        }
+        if let Some(doc) = &self.env.doc(None).ok() {
+            pwd.push_str(&format!(
+                "{}{} ",
+                "doc:".blue(),
+                fmt_short(doc.as_bytes()).blue().bold(),
+            ));
+        }
+        if !pwd.is_empty() {
+            pwd.push('\n');
+        }
+        format!("\n{pwd}{}", "> ".blue())
+    }
+}
+
+#[derive(Debug, Parser)]
+pub enum ReplCmd {
+    #[clap(flatten)]
+    Rpc(#[clap(subcommand)] RpcCommands),
+    /// Quit the Iroh console
+    #[clap(alias = "quit")]
+    Exit,
+}
+
+fn try_parse_cmd<C: Subcommand>(s: &str) -> anyhow::Result<C> {
```

Comment:
Can this handle sth like `doc set my_key hello world`? or at least `doc set my_key "hello world"`?

Reply:
the latter should work, the former not. feel invited to try out some combinations. it uses the [shell_words](https://docs.rs/shell-words/latest/shell_words/) crate to split / escape the command line.